desc/integrals/bounce_integral.py

```diff
@@ -1,26 +1,818 @@
+        Bref : float
+            Optional. Reference magnetic field strength for normalization.
+        Lref : float
+            Optional. Reference length scale for normalization.
+        is_reshaped : bool
+            Whether the arrays in ``data`` are already reshaped to the expected form of
+            shape (..., m, n) or (L, m, n). This option can be used to iteratively
+            compute bounce integrals one flux surface at a time, reducing memory usage
+            To do so, set to true and provide only those axes of the reshaped data.
+            Default is false.
+        check : bool
+            Flag for debugging. Must be false for JAX transformations.
+
+        """
+        errorif(grid.sym, NotImplementedError, msg="Need grid that works with FFTs.")
+        # Strictly increasing zeta knots enforces dζ > 0.
+        # To retain dℓ = (|B|/B^ζ) dζ > 0 after fixing dζ > 0, we require
+        # B^ζ = B⋅∇ζ > 0. This is equivalent to changing the sign of ∇ζ or [∂ℓ/∂ζ]|ρ,a.
+        # Recall dζ = ∇ζ⋅dR, implying 1 = ∇ζ⋅(e_ζ|ρ,a). Hence, a sign change in ∇ζ
+        # requires the same sign change in e_ζ|ρ,a to retain the metric identity.
+        warnif(
+            check and kwargs.pop("warn", True) and jnp.any(data["B^zeta"] <= 0),
+            msg="(∂ℓ/∂ζ)|ρ,a > 0 is required. Enforcing positive B^ζ.",
+        )
+        self._alpha = alpha
+        self._m = grid.num_theta
+        self._n = grid.num_zeta
+        self._x, self._w = get_quadrature(quad, automorphism)
+
+        # peel off field lines
+        iota = data["iota"].ravel()
+        alpha = get_alpha(
+            alpha,
+            iota=(
+                grid.compress(iota)
+                if iota.size == grid.num_nodes
+                # assume passed in reshaped data over flux surface
+                else jnp.array(iota[0])
+            ),
+            num_transit=num_transit,
+            period=2 * jnp.pi,
+        )
+        # Compute spectral coefficients.
+        self._T, self._B = _transform_to_clebsch_1d(
+            grid, alpha, theta, data["|B|"] / Bref, N_B, is_reshaped
+        )
+        self._b_sup_z = _transform_to_desc(
+            grid, jnp.abs(data["B^zeta"]) / data["|B|"] * Lref, is_reshaped
+        )
+        assert self._T.M == self._B.M == num_transit
+        assert self._T.N == theta.shape[-1]
+        assert self._B.N == N_B
+
+    # The Fourier series converges fast for |B|(α, ζ);
+    # for near omnigenous configurations in particular, (∂|B|/∂α)|ρ,ζ vanishes.
+    # However, the Fourier series of θ(α, ζ) converges slower than desired.
+    # Small discontinuities of quantities evaluated between adjacent cuts of a
+    # field line vanish at M > 256. Memory seems much better than DESC's transforms
+    # though, and we avoid creation of a 1000 intermediate dependency arrays of
+    # large size like by avoiding source grid requirement in desc.compute.
+    # edit: this can be reduced once the single valued thing is fixed.
```

Comment:
feel free to ignore this until JAX fixes their fft and dct accuracy. that might have been the issue